apps/docs/pages/guides/auth/auth-hooks.mdx

```diff
@@ -0,0 +1,717 @@
+    from public.password_sign_in_attempts a
+    join auth.users u on a.user_id = u.id
+    where a.user_id = user_id and attempt_successful = false and last_attempt_at > (now() - interval '1 day');
+
+    -- Notify user if the number of failed attempts exceeds the threshold
+    if attempt_count >= max_attempts then
+      -- Fetch the server token
+      select decrypted_secret into server_token from vault.decrypted_secrets where name = 'my_api_key_name';
+
+      -- Prepare the email body
+      email_body := format('{
+        "from": "yoursenderemail@example.com",
+        "to": "%s",
+        "subject": "Security Alert: Repeated Login Attempts Detected",
+        "textbody": "We have detected repeated login attempts for your account. If this was not you, please secure your account.",
+        "htmlbody": "<html><body><strong>Security Alert:</strong> We have detected repeated login attempts for your account. If this was not you, please secure your account.</body></html>",
+        "messagestream": "outbound"
+      }', user_email_address)::jsonb;
+
+      -- Perform the HTTP POST request using Postmark
+      select id into response_id from net.http_post(
+        'https://api.youremailprovider.com/email',
+        email_body,
+        'application/json',
+        array['Accept: application/json', 'X-Postmark-Server-Token: ' || server_token]
+      );
+
+      -- Fetch the response from net._http_response using the obtained id
+      select status_code, content into http_code, error_message from net._http_response where id = response_id;
+
+      -- Handle email sending errors
+      if http_code is null or (http_code < 200 or http_code >= 300) then
+        return jsonb_build_object(
+          'error', jsonb_build_object(
+            'http_code', coalesce(http_code, 0),
+            'message', coalesce(error_message ->> 'message', 'error sending email')
+          )
+        );
+      end if;
+    end if;
+
+    -- Continue with default behavior
+    return jsonb_build_object('decision', 'continue');
+  end;
+$$;
+
+-- Assign appropriate permissions
+grant execute
+  on function public.hook_notify_user_on_failed_attempts
+  to supabase_auth_admin;
+
+revoke execute
+  on function public.hook_notify_user_on_failed_attempts
+  from authenticated, anon;
+
+grant all
+  on table public.password_sign_in_attempts
+  to supabase_auth_admin;
+
+revoke all
+  on table public.password_sign_in_attempts
+  from authenticated, anon;
+```
+
+</TabPanel>
+</Tabs>
+
+## Hook: Custom Access Token
+
+The custom access token hook runs before a token is issued and allows you to add additional claims based on the authentication method used.
+
+Claims returned must conform to our specification. Supabase Auth will check for these claims after the hook is run and return an error if they are not present.
+
+These are the fields currently available on an access token:
+
+| Field           | Required |
```

Comment:
This table renders really terribly. Maybe you can convert these into two lists:

Example:

- Required claims: `aud`, ...
- Optional claims: `jti`, ...